src/internal/scheduler/VirtualTimeScheduler.ts

```diff
@@ -25,7 +25,10 @@ export class VirtualTimeScheduler extends AsyncScheduler {
     const {actions, maxFrames} = this;
     let error: any, action: AsyncAction<any>;
 
-    while ((action = actions.shift()) && (this.frame = action.delay) <= maxFrames) {
+    while ((action = actions[0]) && action.delay <= maxFrames) {
```

Comment:
Seems like this should do the trick:

```
while ((action = actions.shift()) && action.delay <= maxFrames) {
   this.frame = action.delay;
```

Reply:
nope, this changes nothing.
the problem was in `.shift()` mutating the `actions` array **before** actual checks

Reply:
I don't see the issue you're describing. I know the array is being mutated, but you've moved the mutation one line, and there's no checks to the array in that space.

```ts
const arr = [1, 2, 3, 4, 5];
while(arr.shift()) {
  console.log(arr);
}
```

has the same output as

```ts
const arr = [1, 2, 3, 4, 5];
while (arr.shift()) {
  console.log(arr);
}
```

Reply:
I see. You're right, I missed that. Thank you.